src/main/java/net/minestom/server/network/packet/server/play/ParticlePacket.java

```diff
@@ -4,20 +4,56 @@
 import net.minestom.server.network.NetworkBuffer;
 import net.minestom.server.network.packet.server.ServerPacket;
 import net.minestom.server.network.packet.server.ServerPacketIdentifier;
+import net.minestom.server.particle.data.ParticleData;
 import net.minestom.server.utils.PacketUtils;
 import org.jetbrains.annotations.NotNull;
 
+import java.util.Objects;
+
 import static net.minestom.server.network.NetworkBuffer.*;
 
-public record ParticlePacket(int particleId, boolean longDistance,
-                             double x, double y, double z,
-                             float offsetX, float offsetY, float offsetZ,
-                             float particleData, int particleCount, byte[] data) implements ServerPacket {
+public final class ParticlePacket implements ServerPacket {
```

Comment:
Why is this converted away from a record? it should go back

Reply:
we need the particleId on line 46 and 56, i'm not sure using a record is possible here

Reply:
Use the pattern of having a copy constructor and a private static read function, see UnlockRecipesPacket for a similar example